### ISSUE TITLE
Is (or will) it be possible to use ibisml within sklearn pipelines?

### ISSUE THREAD

**Issue by MaxPowerWasTaken** [ibis-project/ibisml#10]
Very happy to find this project, thanks for creating this library! Basically I'm looking for a way to write something resembling a scikit-learn pipeline to train an ML model which can be executed on BQ (even if maybe just partly). I'm also using BQML for some use cases, but it has some limitations that I'm currently jumping back into python for, which has some barriers/constraints in my current context.

From the ibisml readme, it looks to me like the difference between an `ibisml recipe` and a `sklearn pipeline` is that the ibisml recipe only encompasses feature transformations, while an sklearn pipeline includes feature transformations and a model at the end. Is that a fair characterization?

From my perspective, what makes sklearn pipelines so useful is that they encompass both the feature transformations and the model, which can be jointly `fit` on a training dataset, cross-validated on an eval dataset, then re`fit` on a full dataset and used to `predict` on new data. That whole workflow seems to get more complicated if I have to manage separate objects for my feature-transform `recipe` and model.

But maybe it should be possible to write a sklearn pipeline which contains both ibisml feature transformations, and python models at the end? In that case, if the compute for the feature transformations are still farmed out to BQ and the underlying python environment/compute-resource only has to run the `clf.fit` or `clf.predict` part, that would still be very exciting to me.

Thanks again!

**Comment by gforsyth** [comment]
Hey @MaxPowerWasTaken -- welcome!

First, for the quick answer:

> From the ibisml readme, it looks to me like the difference between an `ibisml recipe` and a `sklearn pipeline` is that the ibisml recipe only encompasses feature transformations, while an sklearn pipeline includes feature transformations and a model at the end. Is that a fair characterization?

Yes, that is a fair characterization.  We're currently only covering the "preprocessing" step of an ML pipeline (although that word gets thrown around a lot and has variable definitions).

Now, for the rest of it:

It's not currently possible to use `ibisml` within a `sklearn pipeline` (I don't think).

And I agree that `sklearn pipeline`s are incredibly useful.  Currently, though, they make an assumption that each step of the pipeline is feeding some `X` and some `y` (plus maybe extra kwargs) into the following step (or even looping back for something like N-fold-cross-validation).  In the current paradigm, where the assumptions are that the data look like numpy arrays and pandas dataframes (and that they are split into two separate arrays/dataframes for `X`, `y`) is going to make using ibisml within a sklearn pipeline tricky and/or not that performant.


I think that, broadly, we're thinking about how machine learning pre-processing fits in to a composable data science platform, where you might have several disparate systems tied together with a job runner and ideally using Arrow to shuffle data between the steps of a model training pipeline.  


> But maybe it should be possible to write a sklearn pipeline which contains both ibisml feature transformations, and python models at the end? In that case, if the compute for the feature transformations are still farmed out to BQ and the underlying python environment/compute-resource only has to run the `clf.fit` or `clf.predict` part, that would still be very exciting to me.

I think that this could be possible (mostly) in the future -- if `sklearn` pipelines would accept pyarrow record batches as the inputs to a `sklearn` pipeline, then you could handle the preprocessing and cleaning in BigQuery, then grab those results as `RecordBatchReader`s to be fed into a `sklearn` pipeline to handle training and scoring.

You can try something similar to that today, by exporting the results of an `ibisml` run on BigQuery via `to_pandas`, and then splitting out the `X` and `y` and passing them into your `fit` call.


I'm probably oversimplifying a bunch of ideas here (and missing loads of others) but happy to continue this conversation.  And please do let us know if you try using `ibisml` and where the pain-points are!

**Comment by deepyaman** [comment]
Hi, @MaxPowerWasTaken! I just came across this issue, and wanted to point out that what you've requested is now possible on `main`, as of a [major API refactor](https://github.com/ibis-project/ibisml/pull/25). Thanks for your feedback, and feel free to reopen if there's something I misunderstood that's not addressed.

**Comment by MaxPowerWasTaken** [comment]
Wow this is huge, thanks for tagging me @deepyaman and to the whole team for the work!
